integrators/explicit_embedded_runge_kutta_nyström_integrator_body.hpp

```diff
@@ -0,0 +1,204 @@
+﻿#pragma once
+
+#include "integrators/explicit_embedded_runge_kutta_nyström_integrator.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <ctime>
+#include <vector>
+
+#include "glog/logging.h"
+#include "quantities/quantities.hpp"
+
+namespace principia {
+
+using quantities::Difference;
+using quantities::Quotient;
+
+namespace integrators {
+
+inline ExplicitEmbeddedRungeKuttaNyströmIntegrator const&
+DormandElMikkawyPrince1986RKN434FM() {
+  static ExplicitEmbeddedRungeKuttaNyströmIntegrator const integrator(
+      // c
+      {0.0, 1.0 / 4.0, 7.0 / 10.0, 1.0},
+      // a
+      {{},
+       {1.0 /   32.0},
+       {7.0 / 1000.0, 119.0 / 500.0},
+       {1.0 /   14.0,   8.0 /  27.0, 25.0 / 189.0}},
+      // b̂
+      { 1.0 /  14.0,   8.0 /  27.0,  25.0 / 189.0,  0.0},
+      // b̂′
+      { 1.0 /  14.0,  32.0 /  81.0, 250.0 / 567.0,  5.0 / 54.0},
+      // b
+      {-7.0 / 150.0,  67.0 / 150.0,   3.0 /  20.0, -1.0 / 20.0},
+      // b′
+      {13.0 /  21.0, -20.0 /  27.0, 275.0 / 189.0, -1.0 /  3.0},
+      3);  // lower_order
+  return integrator;
+}
+
+inline ExplicitEmbeddedRungeKuttaNyströmIntegrator::
+           ExplicitEmbeddedRungeKuttaNyströmIntegrator(
+    std::vector<double> const& c,
+    std::vector<std::vector<double>> const& a,
+    std::vector<double> const& b_hat,
+    std::vector<double> const& b_prime_hat,
+    std::vector<double> const& b,
+    std::vector<double> const& b_prime,
+    int const lower_order)
+    : stages_(c.size()),
+      lower_order_(lower_order),
+      c_(c),
+      a_(a),
+      b_hat_(b_hat),
+      b_prime_hat_(b_prime_hat),
+      b_(b),
+      b_prime_(b_prime) {
+  CHECK_EQ(a_.size(), stages_);
+  for(int i = 0; i < stages_; ++i) {
+    CHECK_EQ(a_[i].size(), i);
+  }
+  CHECK_EQ(b_hat_.size(), stages_);
+  CHECK_EQ(b_prime_hat_.size(), stages_);
+  CHECK_EQ(b_.size(), stages_);
+  CHECK_EQ(b_prime_.size(), stages_);
+}
+
+template<typename Position>
+void ExplicitEmbeddedRungeKuttaNyströmIntegrator::Solve(
+    RightHandSideComputation<Position> compute_acceleration,
+    SystemState<Position, Variation<Position>> const& initial_value,
+    Time const& t_final,
+    Time const& first_time_step,
+    StepSizeController<Position> step_size_controller,
+    double const safety_factor,
+    not_null<Solution<Position, Variation<Position>>*> const solution) const {
+  using Displacement = Difference<Position>;
+  using Velocity = Variation<Position>;
+  using Acceleration = Variation<Velocity>;
+
+  // Argument checks.
+  int const dimension = initial_value.positions.size();
+  CHECK_EQ(initial_value.momenta.size(), dimension);
+  CHECK_NE(first_time_step, Time());
+  bool const forward = first_time_step > Time();
```

Comment:
Would a `geometry::Sign integration_direction` work? that way the test for overshooting is much nicer.